packages/with_bloc/lib/src/with_bloc.dart

```diff
@@ -55,26 +59,37 @@ class WithBlocState<BlocType extends ValueNotifier<StateType>, StateType>
     super.didUpdateWidget(oldWidget);
 
     if (!listId.equals(oldWidget.inputs, widget.inputs)) {
-      setState(() {
-        bloc = widget.createBloc(context);
-      });
+      /// Save the previous bloc so we can properly dispose it
+      _oldBloc = bloc;
+
+      /// Recreate the bloc
+      bloc = widget.createBloc(context);
     }
   }
 
   @override
   void dispose() {
-    super.dispose();
-
     bloc.dispose();
+
+    super.dispose();
   }
 
   @override
   Widget build(BuildContext context) {
     return ValueListenableBuilder<StateType>(
       valueListenable: bloc,
       child: widget.child,
-      builder: (context, value, child) =>
-          widget.builder(context, bloc, value, child),
+      builder: (context, value, child) {
+        /// We need to dispose the old bloc
+        if (_oldBloc != null) {
+          /// The old bloc needs to be disposed after [ValueListenableBuilder] had a change to update itself
```

Comment:
@HenriBeck Yeah it kinda looks hacky is there any performance or safety gain with the fix and didn't quite get the issue with "What is tricky here is that we can't just call dispose on it because then the removeListener in the ValueListenableBuilder would fail." can you elaborate a bit more?

Reply:
@Mrtcndkn The issue with disposing the old bloc when creating the new bloc (in `didUpdateWidget`) would be that the `ValueListenableBuilder` will try to unsubscribe (`removeListener`) when it is updated and sees that the bloc reference has changed.

But by that point the old bloc would've already been disposed in which case `removeListener` would throw an error. That's why it's currently inside the inner `Builder`, which executes after the `ValueListenableBuilder` has unsubscribed and hence won't create any issue.

See https://github.com/flutter/flutter/blob/1aedbb1835bd6eb44550293d57d4d124f19901f0/packages/flutter/lib/src/widgets/value_listenable_builder.dart#L166